# [E02] Project 3


> **목차**
===
 - 개요
 - 루브릭 평가기준
 - 용어 정리 및 모델 요약정리
 - Project - 1. 손글씨 분류
 - Project - 2. 와인 분류
 - Project - 3. 유방암 여부 진단
 - 참조
---

> **개요**  
===
1. 어떤 지표가 해석에 주요한 점인지 파악할 것이다.
2. 어떤 모델로 학습하는게 가장 우수한지 비교해볼 것이다.
3. 각 프로젝트마다 오차행렬에서 recall과 precision 중 중점적으로 해석해야 될지를 정해야 한다.

---

> **루브릭 평가기준**
===

## 평가문항 

1. 3가지 데이터 셋의 구성이 합리적으로 진행되었는가?
2. 3가지 데이터 셋에 대해 각각 5가지 모델을 성공적으로 적용하였는가?
3. 3가지 데이터 셋에 대해 모델의 평가지표가 적절히 선택되었는가?

----
## 상세기준

1. Feature와 Label의 선정을 위한 데이터 분석과정이 체계적으로 전개됨
2. 모델 학습 및 테스트가 정상적으로 수행되었음
3. 평가지표 선택 및 이유 설명이 타당함

> **용어 정리 및 모델 요약정리**
-

**지도학습**   

정답이 있는 문제에 대해 학습하는 것을 의미한다.
크게 분류해서 분류와 회귀로 나눠진다.

---
**분류**  
입력받은 데이터를 특정 카테고리 중 하나로 분류해내는 문제
Ex) 아이리스 품종 분류 문제

---
**회귀**  
회귀는 입력받은 데이터에 따라 특정 필드의 수치를 맞히는 문제
Ex) 집에대한 정보(평수, 위치, 층수) 등을 입력받아 가격을 추측하는 문제

---
**비지도 학습**

타깃 데이터가 없어서 무엇을 예측한다기보단 입력데이터에서 어떤 특징을 찾는데 주로 이용된다.

---
**앙상블(Ensemble method)**  
의견을 통합하거나 아ㅕ러가지 결과를 합치는 방식을 데이터 사이언스에서는 앙상블이라 한다.

> **모델 요약 정리**

**[Decision Tree]**  
의사결정나무 라고도 하며 분기에 따라 선택지가 나뉘는 모습이 나무를 거꾸로 담은 것과 같아서 이름 붙여졌다.  

---
장점
+ 해석의 용이성 -> 쉽게 이해가 가능하다. 
+ 교호효과의 해석 -> 두개 이상의 변수가 결합하여 목표 변수에 어떻게 영향을 주는지 쉽게 알 수 있다.
+ 비모수적 모형 -> 선형성, 정규성, 등분산성 등의 가정이 필요없다.

---
단점
+ 비연속성 -> 연속형 변수를 비연속적으로 해석하기 때문에 경계점 부근에서 예측오류가 생길 수 있다.
+ 선형성 or 주효과의 결여
+ 비안정성 -> 분석용 자료에만 의존하므로 새로운 자료의 예측에서는 불안정할 가능성이 높다.

**[Random Forest]**

수많은 의사결정트리가 모여서 만들어진 모델이다.

---
장점  
+ Classification 및 Regression 문제에 모두 적용가능
+ Missing value(결측치)를 다루기 쉬움
+ 대용량 데이터 처리에 효과적
+ 모델의 노이즈를 심화시키는 오버피팅 문제를 회피할 수 있다.
+ Classification 모델에서 상대적으로 중요한 변수 선정 및 Ranking 가능

---
단점  
+ 데이터 수가 많아지면 속도가 크게 느려진다.
+ 결과에 대한 해석이 어렵다.

**[SVM(Support Vector Machine]**  

Support Vector와 Hyperplane(초평면)을 이용해서 분류를 수행

---
**분류방법**  
Decision Boundary와 Support Vector사이의 거리 값을 최대화 시켜서 분류 한다.

---
장점  
+ 범주나 수치 예측 문제에 사용 가능
+ 오류 데이터에 대한 영향이 없다.
+ 과적합이 되는 경우가 적고 
+ 신경망보다 사용하기 쉽다.

---
단점  
+ 여러 개의 조합 테스트 필요
+ 학습 속도가 느림
+ 해석이 어렵고 복잡한 블랙박스 형태로 되어있다.


**[SGD Classifier]**  

Stochastic Gradient Descent Classifier로 배치 크기가 1인 경사하강법 알고리즘이다.  
데이터 셋에서 무작위로 균일하게 선택한 하나의 예를 의존하여 각 단계의 예측 경사를 계산한다.

---
장점  
- 학습 속도가 빠르다.
- 하나의 샘플만 메모리에 있으면 되기 때문에 큰 데이터 셋도 학습 가능

---
단점  
- 샘플 선택이 확률적이기 때문에 배치 경사 하강법에 비해 불안정하다.
- 반복이 충분하면 SGD가 효과적이지만 노이즈가 매우 심하다.
- SGD의 여러 변형 함수의 최저점에 가까운 점을 찾을 가능성이 높지만 항상 보장되지는 않는다.

**[Logistic Regression]**  

소프트맥스 함수를 사용한 다중 클래스 분류 알고리즘이다.  
다중 클래스 분류를 위한 로지스틱 회귀를 소프트 맥스 회귀라고도 한다.  

---
장점  


> **Project - 1 손글씨 분류**
===

> **1) 학습전 추론**
-

불러들인 데이터 값이 전처리가 필요할지 여부에 대해 정확도가 많이 떨어진다면 시행해야겠지만,  
기존 데이터가 이미 가공된 데이터기 때문에 전처리 없이도 정확도에 큰 영향은 없을것이라 생각된다.

손글씨 데이터는 숫자가 맞는지를 판별하는 데이터로 잘못 해석될 시 발생하는 영향이 미미하다.  
때문에 오차행렬을 통한 분석은 필요하지 않다고 여겨진다.  

가장 주의해야 될 점은 데이터값이 치우쳐져 있는지 해석해봐야 한다.  
***why?***   
데이터가 치우쳐져 있으면 오차행렬을 사용하지 않으려고 생각한 만큼 정확도 지표의 신뢰도가 떨어지기 때문이다.  
결론적으로 **정확도**를 가장 중요한 데이터로 여기고 시험할 것이다.

---

> 2) **데이터 준비**  
-

In [1]:
# (1) 필요한 모듈 import
from sklearn.datasets import load_digits # 손글씨 데이터 import
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report, confusion_matrix, classification_report # 오차행렬을 위한 import
from sklearn.tree import DecisionTreeClassifier # 의사결정나무 import
from sklearn.ensemble import RandomForestClassifier  # Random Forest import
from sklearn import svm # svm import
from sklearn.linear_model import SGDClassifier # sgd classifier import
from sklearn.linear_model import LogisticRegression # Logistic Regression import


In [2]:
# (2) 데이터 준비
digits = load_digits()  # 로드한 digits  digits에 넣기
digits_data = digits.data # Feature Data 지정
digit_label = digits.target # Label Data 지정

In [3]:
print(digits.target_names) # 8*8 픽셀로 나눠진걸 볼 수 있다.
print(digits.DESCR)

[0 1 2 3 4 5 6 7 8 9]
.. _digits_dataset:

Optical recognition of handwritten digits dataset
--------------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 1797
    :Number of Attributes: 64
    :Attribute Information: 8x8 image of integer pixels in the range 0..16.
    :Missing Attribute Values: None
    :Creator: E. Alpaydin (alpaydin '@' boun.edu.tr)
    :Date: July; 1998

This is a copy of the test set of the UCI ML hand-written digits datasets
https://archive.ics.uci.edu/ml/datasets/Optical+Recognition+of+Handwritten+Digits

The data set contains images of hand-written digits: 10 classes where
each class refers to a digit.

Preprocessing programs made available by NIST were used to extract
normalized bitmaps of handwritten digits from a preprinted form. From a
total of 43 people, 30 contributed to the training set and different 13
to the test set. 32x32 bitmaps are divided into nonoverlapping blocks of
4x4 and the number of on pixels a

In [4]:

# (3) train, test 데이터 분리
X_train, X_test, y_train, y_test = train_test_split(digits_data, 
                                                    digit_label, 
                                                    test_size=0.2, # 20% 시험 데이터 추출
                                                    random_state=7)

> **3) 다양한 모델로 학습**
-

In [5]:
# Decision Tree
decision_tree = DecisionTreeClassifier()
decision_tree.fit(X_train, y_train)
y_pred = decision_tree.predict(X_test)

print(classification_report(y_test, y_pred))


              precision    recall  f1-score   support

           0       0.96      1.00      0.98        43
           1       0.83      0.81      0.82        42
           2       0.82      0.82      0.82        40
           3       0.91      0.91      0.91        34
           4       0.83      0.92      0.87        37
           5       0.87      0.96      0.92        28
           6       0.93      0.93      0.93        28
           7       0.87      0.82      0.84        33
           8       0.82      0.65      0.73        43
           9       0.74      0.81      0.78        32

    accuracy                           0.86       360
   macro avg       0.86      0.86      0.86       360
weighted avg       0.86      0.86      0.86       360



In [6]:
# Random Forest
randomforest = RandomForestClassifier()
randomforest.fit(X_train, y_train)
y_pred = randomforest.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.98      0.99        43
           1       0.93      1.00      0.97        42
           2       1.00      1.00      1.00        40
           3       0.92      1.00      0.96        34
           4       0.93      1.00      0.96        37
           5       0.93      1.00      0.97        28
           6       1.00      0.93      0.96        28
           7       0.94      1.00      0.97        33
           8       1.00      0.84      0.91        43
           9       1.00      0.91      0.95        32

    accuracy                           0.96       360
   macro avg       0.97      0.96      0.96       360
weighted avg       0.97      0.96      0.96       360



In [7]:
# SVM
svm_model = svm.SVC()

svm_model.fit(X_train, y_train)
y_pred = svm_model.predict(X_test)

print(classification_report(y_test, y_pred)) 

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        43
           1       0.95      1.00      0.98        42
           2       1.00      1.00      1.00        40
           3       1.00      1.00      1.00        34
           4       1.00      1.00      1.00        37
           5       0.93      1.00      0.97        28
           6       1.00      1.00      1.00        28
           7       1.00      1.00      1.00        33
           8       1.00      0.93      0.96        43
           9       1.00      0.97      0.98        32

    accuracy                           0.99       360
   macro avg       0.99      0.99      0.99       360
weighted avg       0.99      0.99      0.99       360



In [8]:
# SGD Classifier
sgd_model = SGDClassifier()

sgd_model.fit(X_train, y_train)
y_pred = sgd_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.98      0.99        43
           1       0.92      0.86      0.89        42
           2       0.95      1.00      0.98        40
           3       0.94      0.91      0.93        34
           4       0.93      1.00      0.96        37
           5       0.93      1.00      0.97        28
           6       1.00      0.96      0.98        28
           7       0.97      0.97      0.97        33
           8       0.90      0.88      0.89        43
           9       0.94      0.94      0.94        32

    accuracy                           0.95       360
   macro avg       0.95      0.95      0.95       360
weighted avg       0.95      0.95      0.95       360



In [9]:
# Logistic Regression
logistic_model = LogisticRegression(max_iter = 2500) # 그냥 코드로 돌렸을 시에 정답에 도달하기전에 종료되서 max iter를 늘려줬다.


logistic_model.fit(X_train, y_train)
y_pred = logistic_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        43
           1       0.95      0.95      0.95        42
           2       0.98      1.00      0.99        40
           3       0.94      0.97      0.96        34
           4       1.00      1.00      1.00        37
           5       0.79      0.96      0.87        28
           6       1.00      0.96      0.98        28
           7       0.94      0.97      0.96        33
           8       0.92      0.81      0.86        43
           9       0.97      0.88      0.92        32

    accuracy                           0.95       360
   macro avg       0.95      0.95      0.95       360
weighted avg       0.95      0.95      0.95       360



> **4) 결론**
-

## Data Base
훈련 data : 80%  
시험 data : 10% 

---
## Decision Tree 

정확도 : 85%        

---   
## Random Forest

정확도 : 95%

---
## SVM 채택

정확도 : 99%

---
## SGD Classifier

정확도 : 95%

---
## Logistic Regression

정확도 : 95%

---

> **회고**
-

사전에 전처리의 필요성에 대해 생각해봤을때 가공된 데이터기 때문에 필요없을 것이라 예상한점은 맞았다.  
데이터 값이 맞고 틀림에 따른 인과관계에 대한 관점 없이 정확도만으로 판단했을때 SVM 모델로 학습했을때 가장 우수한 모델이 나왔다.  
그 외 나머지는 Decision Tree 를 제외하고는 95%이상의 값이 나왔는데, Random Forest 모델이 Decision Tree의 상위 버전인 만큼 정확도가  
더 좋게 나옴을 볼 수 있었다.  

결론적으로는 사전 예측때 그랬던 것처럼 정확도를 기준으로 판별했을 때 SVM이 가장 높았기 때문에  
이 손글씨 분류에서는 SVM 모델이 가장 적합하다고 생각된다.

---
> **어려웠던 점**
-

Logistic Regression 에서 디폴트값으로 연산하면 계속해서 오류가 발생했다.  
이유는 값에 가까이는 가도 도달하지 못한채로 연산이 종료 됐다.  
이 문제를 해결하기 위해서 구글링을 통해서 max_iter 값을 늘려주면 된다는걸 깨달았고, 
값을 계속해서 올려주면서 적절한 반복횟수를 찾아 낼 수 있었다.

---

> **Project -2 와인 분류**
===

> **1) 학습전 추론**
-

Project - 1 에서도 그랬지만 CSV 파일은 정형 데이터기 때문에 별도의 전처리 작업이 없어도 정확도가 나옴을 알 수 있다.  
와인을 세개로 분류하는 작업을 하는데 손글씨 분류 작업과 마찬가지로 어떤 특정 지표가 잘못 계산됨에 따라 치명적이지 않고,
이진분류가 아니기때문에 오차행렬 분석은 시행하지 않는다.
정확도를 가지고 모델의 성능을 파악할 예정이다.
데이터의 양이 적고 3개의 클래스를 분류하는 것이기 때문에 많은 표본이 필요한 모델들은 정확도가 적게 나올것으로 예상되고  
의사결정나무나 Random Forest 등이 좀 더 적합한 모델일 것으로 추측된다.

> **2) 데이터 준비**
-

In [10]:
# (1) 필요한 모듈 import
from sklearn.datasets import load_wine # 와인 데이터 import
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report, confusion_matrix, classification_report # 오차행렬을 위한 import
from sklearn.tree import DecisionTreeClassifier # 의사결정나무 import
from sklearn.ensemble import RandomForestClassifier  # Random Forest import
from sklearn import svm # svm import
from sklearn.linear_model import SGDClassifier # sgd classifier import
from sklearn.linear_model import LogisticRegression # Logistic Regression import

In [11]:
# (2) 데이터 준비
wine = load_wine()  # 로드한 wine데이터 wine에 넣기
wine_data = wine.data # Feature Data 지정
wine_label = wine.target # Label Data 지정

In [12]:
print(wine.target_names) # 총 세개로 나눠짐 ['class_0' 'class_1' 'class_2']
print(wine.DESCR) # 와인에 관한 개요

['class_0' 'class_1' 'class_2']
.. _wine_dataset:

Wine recognition dataset
------------------------

**Data Set Characteristics:**

    :Number of Instances: 178 (50 in each of three classes)
    :Number of Attributes: 13 numeric, predictive attributes and the class
    :Attribute Information:
 		- Alcohol
 		- Malic acid
 		- Ash
		- Alcalinity of ash  
 		- Magnesium
		- Total phenols
 		- Flavanoids
 		- Nonflavanoid phenols
 		- Proanthocyanins
		- Color intensity
 		- Hue
 		- OD280/OD315 of diluted wines
 		- Proline

    - class:
            - class_0
            - class_1
            - class_2
		
    :Summary Statistics:
    
    ============================= ==== ===== ======= =====
                                   Min   Max   Mean     SD
    ============================= ==== ===== ======= =====
    Alcohol:                      11.0  14.8    13.0   0.8
    Malic Acid:                   0.74  5.80    2.34  1.12
    Ash:                          1.36  3.23    2.36  0.27
   

In [13]:
# (3) train, test 데이터 분리
X_train, X_test, y_train, y_test = train_test_split(wine_data, 
                                                    wine_label, 
                                                    test_size=0.3,# 30% 시험 데이터 추출
                                                    random_state=5)

> **3) 다양한 모델로 학습**
-

In [14]:
# Decision Tree
decision_tree = DecisionTreeClassifier()
decision_tree.fit(X_train, y_train)
y_pred = decision_tree.predict(X_test)

print(classification_report(y_test, y_pred))


              precision    recall  f1-score   support

           0       0.95      0.91      0.93        23
           1       0.89      0.89      0.89        18
           2       0.93      1.00      0.96        13

    accuracy                           0.93        54
   macro avg       0.92      0.93      0.93        54
weighted avg       0.93      0.93      0.93        54



In [15]:
# Random Forest
randomforest = RandomForestClassifier()
randomforest.fit(X_train, y_train)
y_pred = randomforest.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.96      0.98        23
           1       0.94      0.94      0.94        18
           2       0.93      1.00      0.96        13

    accuracy                           0.96        54
   macro avg       0.96      0.97      0.96        54
weighted avg       0.96      0.96      0.96        54



In [16]:
# SVM
svm_model = svm.SVC()

svm_model.fit(X_train, y_train)
y_pred = svm_model.predict(X_test)

print(classification_report(y_test, y_pred)) 

              precision    recall  f1-score   support

           0       0.89      0.74      0.81        23
           1       0.60      0.83      0.70        18
           2       0.40      0.31      0.35        13

    accuracy                           0.67        54
   macro avg       0.63      0.63      0.62        54
weighted avg       0.68      0.67      0.66        54



In [17]:
# SGD Classifier
sgd_model = SGDClassifier()

sgd_model.fit(X_train, y_train)
y_pred = sgd_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.69      0.96      0.80        23
           1       0.59      0.72      0.65        18
           2       0.00      0.00      0.00        13

    accuracy                           0.65        54
   macro avg       0.43      0.56      0.48        54
weighted avg       0.49      0.65      0.56        54



/opt/conda/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1308: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/opt/conda/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1308: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/opt/conda/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1308: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [18]:
# Logistic Regression
logistic_model = LogisticRegression(max_iter = 2500) # 그냥 코드로 돌렸을 시에 정답에 도달하기전에 종료되서 max iter를 늘려줬다.


logistic_model.fit(X_train, y_train)
y_pred = logistic_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.96      0.98        23
           1       0.89      0.94      0.92        18
           2       0.92      0.92      0.92        13

    accuracy                           0.94        54
   macro avg       0.94      0.94      0.94        54
weighted avg       0.95      0.94      0.94        54



> **4) 결론**
-

## Data Base
훈련 data : 70%  
시험 data : 30% 

---
## Decision Tree 

정확도 : 93%        

---   
## Random Forest 채택

정확도 : 98%

---
## SVM

정확도 : 67%

---
## SGD Classifier

정확도 : 57%

---
## Logistic Regression

정확도 : 94%

---

> **회고**
-

손글씨 분류와는 다르게 훈련에 쓰일 데이터 양이 적어서 그런지 SVM이나 SGD Classifier등에서는 정확도가 아주 적게 나옴을 알 수 있다.
처음에는 표본이 적기 때문에 9:1의 비율로 훈련데이터와 시험데이터를 나눴었는데 이렇게 분류하니깐 SGD Classifier에서 분류를 못하는 현상이 발생해서 훈련에 필요한 데이터량을 좀 더 추가 했다.
적은 양의 데이터에서는 Random Forest가 아주 적합한 모델임을 알 수 있었고, 표본의 수를 늘리는게 정확도 계산에 중요한 역할을 한다는걸 깨달을 수 있었다.

결론적으로는 와인 분류에서는 Random Forest 모델이 가장 적합하다고 생각된다.

---
> **어려웠던 점**
-

회고에도 썼지만 SGD Classifier가 시험 표본이 적을 경우에는 제대로 나눠지질 않아서 고민을 많이 하고 구글링을 하면서 공부했다.
그 외에는 손글씨랑 거의 유사하기 때문에 어려운 점은 없었다.

---

> **Project -3 유방암 여부 진단**
-

> **1) 학습전 추론**
-

다른 두 프로젝트와 마찬가지로 가공된 CVS 파일 이기 때문에 별도의 전처리는 없이 훈련 및 시험을 하지만  
다른 점은 암진단여부를 가리기 때문에 이진행렬이기도 하고 오진으로 인한 위험을 줄이기 위해 오차행렬분석도 같이 병행할 것이다.
단순 정확도 추정만이 아니기 때문에 복합적으로 판단해야해서 좀 더 난이도가 있을 것으로 예상된다.

오차행렬분석에서 암진단에 있어서 암환자인데 건강하다고 오진할 확률이 적을 때 이상적인 수치가 만들어질것이다.



> **2) 데이터 준비**
-

In [19]:
# (1) 필요한 모듈 import
from sklearn.datasets import load_breast_cancer # 유방암진단 데이터 import
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report, confusion_matrix, classification_report # 오차행렬을 위한 import
from sklearn.tree import DecisionTreeClassifier # 의사결정나무 import
from sklearn.ensemble import RandomForestClassifier  # Random Forest import
from sklearn import svm # svm import
from sklearn.linear_model import SGDClassifier # sgd classifier import
from sklearn.linear_model import LogisticRegression # Logistic Regression import

In [20]:
# (2) 데이터 준비
breast_cancer = load_breast_cancer()  # 로드한 유방암 진단 데이터 breast_cancer에 넣기
breast_cancer_data = breast_cancer.data # Feature Data 지정
breast_cancer_label = breast_cancer.target # Label Data 지정

In [21]:
print(breast_cancer.target_names) # 두 개로 나눠짐 ['malignant' 'benign']
print(breast_cancer.DESCR) # 유방암 진단에 관한 개요

['malignant' 'benign']
.. _breast_cancer_dataset:

Breast cancer wisconsin (diagnostic) dataset
--------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 569

    :Number of Attributes: 30 numeric, predictive attributes and the class

    :Attribute Information:
        - radius (mean of distances from center to points on the perimeter)
        - texture (standard deviation of gray-scale values)
        - perimeter
        - area
        - smoothness (local variation in radius lengths)
        - compactness (perimeter^2 / area - 1.0)
        - concavity (severity of concave portions of the contour)
        - concave points (number of concave portions of the contour)
        - symmetry
        - fractal dimension ("coastline approximation" - 1)

        The mean, standard error, and "worst" or largest (mean of the three
        worst/largest values) of these features were computed for each image,
        resulting in 30 features.  For instanc

In [22]:
# (3) train, test 데이터 분리
X_train, X_test, y_train, y_test = train_test_split(breast_cancer_data, 
                                                    breast_cancer_label, 
                                                    test_size=0.25,# 25% 시험 데이터 추출
                                                    random_state=3)

> **3) 다양한 모델로 학습**
-

In [23]:
# Decision Tree
decision_tree = DecisionTreeClassifier()
decision_tree.fit(X_train, y_train)
y_pred = decision_tree.predict(X_test)

print(classification_report(y_test, y_pred))
print(confusion_matrix(y_test, y_pred)) # 오차행렬 분석을 위한 값 각각 TP,FN,FP,TN 이다. TP, TN이 클수록 좋다.
conf_mat = confusion_matrix(y_test, y_pred)
precision = conf_mat[0,0] / (conf_mat[0,1] + conf_mat[0,0]) # 정밀도
recall =  conf_mat[0,0] / (conf_mat[1,1] + conf_mat[0,0]) # 재현율
print("정밀도 : {}\n재현율 : {}\n".format(precision,recall))

              precision    recall  f1-score   support

           0       0.92      0.91      0.91        53
           1       0.95      0.96      0.95        90

    accuracy                           0.94       143
   macro avg       0.93      0.93      0.93       143
weighted avg       0.94      0.94      0.94       143

[[48  5]
 [ 4 86]]
정밀도 : 0.9056603773584906
재현율 : 0.3582089552238806



In [24]:
# Random Forest
randomforest = RandomForestClassifier()
randomforest.fit(X_train, y_train)
y_pred = randomforest.predict(X_test)

print(classification_report(y_test, y_pred))
print(confusion_matrix(y_test, y_pred)) # 오차행렬 분석을 위한 값 각각 TP,FN,FP,TN 이다. TP, TN이 클수록 좋다.
conf_mat = confusion_matrix(y_test, y_pred)
precision = conf_mat[0,0] / (conf_mat[0,1] + conf_mat[0,0]) # 정밀도
recall =  conf_mat[0,0] / (conf_mat[1,1] + conf_mat[0,0]) # 재현율
print("정밀도 : {}\n재현율 : {}\n".format(precision,recall))

              precision    recall  f1-score   support

           0       0.92      0.91      0.91        53
           1       0.95      0.96      0.95        90

    accuracy                           0.94       143
   macro avg       0.93      0.93      0.93       143
weighted avg       0.94      0.94      0.94       143

[[48  5]
 [ 4 86]]
정밀도 : 0.9056603773584906
재현율 : 0.3582089552238806



In [25]:
# SVM
svm_model = svm.SVC()

svm_model.fit(X_train, y_train)
y_pred = svm_model.predict(X_test)

print(classification_report(y_test, y_pred)) 
print(confusion_matrix(y_test, y_pred)) # 오차행렬 분석을 위한 값 각각 TP,FN,FP,TN 이다. TP, TN이 클수록 좋다.
conf_mat = confusion_matrix(y_test, y_pred)
precision = conf_mat[0,0] / (conf_mat[0,1] + conf_mat[0,0]) # 정밀도
recall =  conf_mat[0,0] / (conf_mat[1,1] + conf_mat[0,0]) # 재현율
print("정밀도 : {}\n재현율 : {}\n".format(precision,recall))

              precision    recall  f1-score   support

           0       0.93      0.79      0.86        53
           1       0.89      0.97      0.93        90

    accuracy                           0.90       143
   macro avg       0.91      0.88      0.89       143
weighted avg       0.90      0.90      0.90       143

[[42 11]
 [ 3 87]]
정밀도 : 0.7924528301886793
재현율 : 0.32558139534883723



In [26]:
# SGD Classifier
sgd_model = SGDClassifier()

sgd_model.fit(X_train, y_train)
y_pred = sgd_model.predict(X_test)

print(classification_report(y_test, y_pred))
print(confusion_matrix(y_test, y_pred)) # 오차행렬 분석을 위한 값 각각 TP,FN,FP,TN 이다. TP, TN이 클수록 좋다.
conf_mat = confusion_matrix(y_test, y_pred)
precision = conf_mat[0,0] / (conf_mat[0,1] + conf_mat[0,0]) # 정밀도
recall =  conf_mat[0,0] / (conf_mat[1,1] + conf_mat[0,0]) # 재현율
print("정밀도 : {}\n재현율 : {}\n".format(precision,recall))

              precision    recall  f1-score   support

           0       0.64      0.96      0.77        53
           1       0.97      0.68      0.80        90

    accuracy                           0.78       143
   macro avg       0.80      0.82      0.78       143
weighted avg       0.85      0.78      0.79       143

[[51  2]
 [29 61]]
정밀도 : 0.9622641509433962
재현율 : 0.45535714285714285



In [27]:
# Logistic Regression
logistic_model = LogisticRegression(max_iter = 2500) # 그냥 코드로 돌렸을 시에 정답에 도달하기전에 종료되서 max iter를 늘려줬다.


logistic_model.fit(X_train, y_train)
y_pred = logistic_model.predict(X_test)

print(classification_report(y_test, y_pred))
print(confusion_matrix(y_test, y_pred)) # 오차행렬 분석을 위한 값 각각 TP,FN,FP,TN 이다. TP, TN이 클수록 좋다.
conf_mat = confusion_matrix(y_test, y_pred)
precision = conf_mat[0,0] / (conf_mat[0,1] + conf_mat[0,0]) # 정밀도
recall =  conf_mat[0,0] / (conf_mat[1,1] + conf_mat[0,0]) # 재현율
print("정밀도 : {}\n재현율 : {}\n".format(precision,recall))

              precision    recall  f1-score   support

           0       0.94      0.92      0.93        53
           1       0.96      0.97      0.96        90

    accuracy                           0.95       143
   macro avg       0.95      0.95      0.95       143
weighted avg       0.95      0.95      0.95       143

[[49  4]
 [ 3 87]]
정밀도 : 0.9245283018867925
재현율 : 0.3602941176470588



> **4) 결론**
-

## Data Base
훈련 data : 75%  
시험 data : 25% 

---
## Decision Tree 채택

정확도 : 94%  
정밀도 : 93%    
재현율 : 37%  

---   
## Random Forest

정확도 : 94%   
정밀도 : 92%   
재현율 : 36%   

---
## SVM 

정확도 : 90%  
정밀도 : 79%   
재현율 : 33%   

---
## SGD Classifier

정확도 : 70% ~ 90%
정밀도 : 70% ~ 100%
재현율 : 33% ~ 53%  

---
## Logistic Regression 

정확도 : 95%  
정밀도 : 92%  
재현율 : 36%  

---

> **회고**
-

가장 중요한 점은 양성인데 음성으로 판단 하는 경우가 적어야 되서 재현율(Recall)을 가장 중요하게 판단하고 그 다음으로 정밀도, 정확도 순으로 점수를 매겨서 채택했다. 의사 결정나무가 가장 간단한 로직을 가지고 있고 그 업그레이드 버전이 random forest라고 생각해서 고민을 했지만 시험 지표상으로 가장 중점적으로 봤던 재현율이 가장 높기도 하고 그 후에 정밀도도 높아서 가장 이상적인 모델로 결정했다.

---
> **어려웠던 점**
-

SGD Classifier는 매커니즘 자체가 확률로 만들어져있기 때문인지 몰라도 재현율이 때로는 가장 이상적으로 나와서 고민을 했지만  
수치 변동이 너무 커서 의사결정나무랑 고민하는데에 시간을 많이 소비했다. 좀 더 표본이 많다면 이렇게까지 변동이 크지 않았을텐데 하는 아쉬움이 남았다.  

재현율이 시험 표본이 적었을 때 너무 적게 나오는 현상이 발생해서 적절한 값을 유도하면서 고민을 했고, 현재도 재현율이 너무 낮게 나와서 이게 맞게 됐는지 걱정스럽다.

---

> 참조
-

1) 의사결정나무 장단점 - https://dreamlog.tistory.com/576  
2) Random Forest 장단점 - https://wooono.tistory.com/115  /  https://jjeongil.tistory.com/908  
3) SVM 장단점 https://muzukphysics.tistory.com/entry/ML-8-%EB%A8%B8%EC%8B%A0%EB%9F%AC%EB%8B%9D-SVM-%EA%B8%B0%EB%B3%B8-%EA%B0%9C%EB%85%90%EA%B3%BC-%EC%9E%A5%EB%8B%A8%EC%A0%90-Support-Vector-Machine  
4) 경사하강법 장단점 : https://velog.io/@arittung/DeepLearningStudyDay8  
5) SGD Classifier 오류 해결 - https://stackoverflow.com/questions/43162506/undefinedmetricwarning-f-score-is-ill-defined-and-being-set-to-0-0-in-labels-wi  
6) 오차행렬분석 - https://kimdingko-world.tistory.com/173



